### PR TITLE
Create the Vnet only and hash the nameservers

### PR DESCRIPTION
**Pull request by asalkeld**
```release-note
NONE
```
Get the nameservers from the vnet and add it to the scaleset hash to cause vm rotate when the nameservers are changed.
/cc @jim-minter

### CODE REVIEW

**Comment by jim-minter** [comment]
I think this is OK but I'm in a rush - @ehashman can you take a look too?

**Comment by mjudeikis** [comment]
lgtm

File: pkg/entrypoint/startup/startup.go [hunk 1]
```diff
@@ -40,6 +42,14 @@ func runStartup(ctx context.Context, log *logrus.Entry) error {
 		return err
 	}
 
+	ns, err := startup.GetNameserversFromDNSConfig("/host")
+	if err != nil {
+		return err
+	}
+	if !reflect.DeepEqual(ns, cs.Properties.NetworkProfile.InUseNameservers) {
+		log.Warnf("overriding InUseNameservers with values from host resolver difference %s", cmp.Diff(ns, cs.Properties.NetworkProfile.InUseNameservers))
```
Comment (asalkeld) [inline pkg/entrypoint/startup/startup.go]: this to me is a bit troubling what to do if the vnet and resolver come up with different values..

File: pkg/plugin/plugin.go [hunk 2]
```diff
@@ -82,6 +92,8 @@ func (p *plugin) ValidateAdmin(ctx context.Context, new, old *api.OpenShiftManag
 	validator := validate.NewAdminValidator(p.testConfig.RunningUnderTest)
 	errs = validator.Validate(new, old)
 
+	// TODO figure out what kind of validation we want here for nameservers.
+
```
Comment (asalkeld) [inline pkg/plugin/plugin.go]: @jim-minter remember this..

File: pkg/startup/v12/data/worker/etc/origin/node/resolv.conf [hunk 3]
```diff
@@ -1 +1,3 @@
-nameserver 168.63.129.16
+{{ range .Properties.NetworkProfile.InUseNameservers }}
```
Comment (asalkeld) [inline pkg/startup/v12/data/worker/etc/origin/node/resolv.conf]: using this InUseNameserver makes life a lot easier

**Comment by asalkeld** [comment]
@jim-minter ptal, I think this sucks less. :thinking:

**Comment by jim-minter** [comment]
unit test failure

File: pkg/api/2019-10-27-preview/types.go [hunk 4]
```diff
@@ -103,6 +103,12 @@ type NetworkProfile struct {
 
 	// VnetID (out): the ID of the Vnet created for the OSA cluster
 	VnetID *string `json:"vnetId,omitempty"`
+
+	// AllowVnetChangesToCauseNodeRotation (in): if the nameservers are changed on
+	// the vnet then allow the update to cause nodes to rotate.
+	// Note: this is just a safety mechanism so this does not happen unintentionally
+	// but this *must* be done for the nodes to use of the new nameservers.
+	AllowVnetChangesToCauseNodeRotation *bool `json:"allowVnetChangesToCauseNodeRotation,omitempty"`
```
Comment (jim-minter) [inline pkg/api/2019-10-27-preview/types.go]: @amanohar need to close on the name of this flag

File: pkg/api/types.go [hunk 5]
```diff
@@ -140,6 +140,14 @@ type NetworkProfile struct {
...
 	// the RP in the deployment callback function
 	InternalLoadBalancerFrontendIPID string `json:"-"`
+	// AllowVnetChangesToCauseNodeRotation (in): if the nameservers are changed on
+	// the vnet then allow the update to cause nodes to rotate.
+	// Note: this is just a safety mechanism so this does not happen unintentionally
+	// but this *must* be done for the nodes to use of the new nameservers.
+	AllowVnetChangesToCauseNodeRotation bool `json:"allowVnetChangesToCauseNodeRotation,omitempty"`
+	// InUseNameservers is the list of DNS nameservers that the running nodes
+	// are using.
+	InUseNameservers []string `json:"inUseNameservers,omitempty"`
```
Comment (jim-minter) [inline pkg/api/types.go]: Why not
```
Nameservers []string `json:"nameservers,omitempty"
```
?

File: pkg/plugin/plugin.go [hunk 6]
```diff
@@ -291,6 +303,13 @@ func (p *plugin) createOrUpdateExt(ctx context.Context, cs *api.OpenShiftManaged
 		cs.Properties.NetworkProfile.InternalLoadBalancerFrontendIPID = resourceid.ResourceID(cs.Properties.AzProfile.SubscriptionID, cs.Properties.AzProfile.ResourceGroup, "Microsoft.Network/loadBalancers", armconst.IlbAPIServerName+"/frontendIPConfigurations/"+armconst.IlbAPIServerFrontendConfigurationName)
 	}
 
+	if !isUpdate || cs.Properties.NetworkProfile.AllowVnetChangesToCauseNodeRotation {
```
Comment (jim-minter) [inline pkg/plugin/plugin.go]: Do we want to override `AllowVnetChangesToCauseNodeRotation` to false at some point to ensure it is a transient flag and not written to the database?

File: pkg/arm/nameservers.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+		return nil, err
+	}
+	vcli := network.NewVirtualNetworkClient(ctx, log, subscriptionID, authorizer)
+	vnet, err := vcli.Get(ctx, resourceGroup, constants.VnetName, "")
+	if err != nil {
+		de, ok := err.(autorest.DetailedError)
+		if ok {
+			re, ok := de.Original.(*azure.RequestError)
+			if ok {
+				if re.ServiceError.Code == "ResourceNotFound" {
```
Comment (jim-minter) [inline pkg/arm/nameservers.go]: I feel like the ResourceNotFound case here should be avoidable if this is called only on update and not create - can this be done?

File: pkg/entrypoint/startup/startup.go (hunk 1 above)
Comment (jim-minter) [reply pkg/entrypoint/startup/startup.go]: yes; I think you're doing the right thing and I think it's the best that can be done.

File: pkg/plugin/plugin.go (hunk 2 above)
Comment (jim-minter) [reply pkg/plugin/plugin.go]: I don't think there's any validation to be done here?

File: pkg/arm/nameservers.go (hunk 7 above)
Comment (asalkeld) [reply pkg/arm/nameservers.go]: I'll have a look and see why i needed this

File: pkg/plugin/plugin.go [hunk 8]
```diff
@@ -127,6 +141,7 @@ func (p *plugin) GenerateConfig(ctx context.Context, cs *api.OpenShiftManagedClu
 	if !isUpdate || cs.Config.PluginVersion == "latest" {
 		cs.Config.PluginVersion = p.pluginConfig.PluginVersion
 		setVersionFields = true
+		cs.Properties.RefreshCluster = true
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/plugin/plugin.go]: ok

File: pkg/plugin/plugin.go [hunk 9]
```diff
@@ -73,6 +73,20 @@ func (p *plugin) Validate(ctx context.Context, new, old *api.OpenShiftManagedClu
 			errs = append(errs, fmt.Errorf(`cluster with version %q cannot be updated by resource provider with version %q`, new.Config.PluginVersion, p.pluginConfig.PluginVersion))
 		}
 	}
+	expectingRotations := (new.Properties.RefreshCluster || new.Properties.ClusterVersion == "latest" && new.Config.PluginVersion == "latest")
+	var ns []string
+	if old != nil && !expectingRotations {
+		clusterUpgrader, err := p.upgraderFactory(ctx, p.log, new, true, true, p.testConfig)
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/plugin/plugin.go]: will do

File: pkg/plugin/plugin.go [hunk 10]
```diff
@@ -291,6 +326,17 @@ func (p *plugin) createOrUpdateExt(ctx context.Context, cs *api.OpenShiftManaged
 		cs.Properties.NetworkProfile.InternalLoadBalancerFrontendIPID = resourceid.ResourceID(cs.Properties.AzProfile.SubscriptionID, cs.Properties.AzProfile.ResourceGroup, "Microsoft.Network/loadBalancers", armconst.IlbAPIServerName+"/frontendIPConfigurations/"+armconst.IlbAPIServerFrontendConfigurationName)
 	}
 
+	if !isUpdate {
+		cs.Properties.NetworkProfile.Nameservers = []string{armconst.AzureNameserver}
+	} else if cs.Properties.RefreshCluster {
+		cs.Properties.NetworkProfile.Nameservers, err = clusterUpgrader.GetNameserversFromVnet(ctx, p.log, cs.Properties.AzProfile.SubscriptionID, cs.Properties.AzProfile.ResourceGroup)
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/plugin/plugin.go]: I'll remove, it is redundant.

File: pkg/plugin/plugin.go [hunk 11]
```diff
@@ -168,6 +183,13 @@ func (p *plugin) ListEtcdBackups(ctx context.Context, cs *api.OpenShiftManagedCl
 
 func (p *plugin) RecoverEtcdCluster(ctx context.Context, cs *api.OpenShiftManagedCluster, deployFn api.DeployFn, backupBlob string) *api.PluginError {
 	suffix := fmt.Sprintf("%d", p.now().Unix())
+	changed, err := p.vnetNameserversHaveChanged(ctx, cs)
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/plugin/plugin.go]: ok, i'll remove

File: pkg/plugin/plugin.go [hunk 12]
```diff
@@ -74,6 +76,12 @@ func (p *plugin) Validate(ctx context.Context, new, old *api.OpenShiftManagedClu
 		}
 	}
 
+	if old != nil && !new.Properties.RefreshCluster {
```
Comment (jim-minter) [inline pkg/plugin/plugin.go]: above this line, suggest returning early if `len(errs) > 0`

File: pkg/plugin/plugin.go [hunk 13]
```diff
@@ -90,6 +98,11 @@ func (p *plugin) ValidateAdmin(ctx context.Context, new, old *api.OpenShiftManag
 			errs = append(errs, fmt.Errorf(`cluster with version %q cannot be updated by resource provider with version %q`, new.Config.PluginVersion, p.pluginConfig.PluginVersion))
 		}
 	}
+	if old != nil && !new.Properties.RefreshCluster {
```
Comment (jim-minter) [inline pkg/plugin/plugin.go]: and here

File: pkg/util/pluginversion/pluginversion.go [hunk 14]
```diff
@@ -0,0 +1,40 @@
+package pluginversion
+
+import "fmt"
+
+// Version represents the plugin version
+type Version struct {
+	Major  int
+	Minor  int
+	Latest bool
```
Comment (jim-minter) [inline pkg/util/pluginversion/pluginversion.go]: yikes - when do we need this case?

**Comment by jim-minter** [comment]
@asalkeld please check test failures

File: pkg/util/pluginversion/pluginversion.go (hunk 14 above)
Comment (asalkeld) [reply pkg/util/pluginversion/pluginversion.go]: Just trying to be complete, i'll remove.

File: pkg/plugin/plugin.go (hunk 12 above)
Comment (asalkeld) [reply pkg/plugin/plugin.go]: just tried, makes the unit tests horrid as the configs have to be perfect to catch the error i want to catch.

File: pkg/plugin/plugin.go [hunk 15]
```diff
@@ -247,6 +265,28 @@ func (p *plugin) CreateOrUpdate(ctx context.Context, cs *api.OpenShiftManagedClu
 	return p.createOrUpdateExt(ctx, cs, updateTypeCreate, deployFn, false)
 }
 
+func (p *plugin) checkIfClusterWillRefresh(ctx context.Context, cs *api.OpenShiftManagedCluster) error {
+	if cs.Properties.RefreshCluster == nil || (cs.Properties.RefreshCluster != nil && *cs.Properties.RefreshCluster) {
```
Comment (jim-minter) [inline pkg/plugin/plugin.go]: nit - this reduces to `if cs.Properties.RefreshCluster == nil || *cs.Properties.RefreshCluster` ("if it's nil, or it's set to true")

**Pull request merged**